lib/boss/participant.rb

```diff
@@ -0,0 +1,357 @@
+    #
+    def message; opt('message') || encode_workitem; end
+
+    # Returns the routing key for the message to publish.
+    #
+    # Available as a method so it can be overriden (the return value could
+    # depend on the @workitem or other factors).
+    #
+    def routing_key; opt('routing_key'); end
+
+    # Returns whether the publish should be persistent or not.
+    #
+    # Available as a method so it can be overriden (the return value could
+    # depend on the @workitem or other factors).
+    #
+    def persistent; opt('persistent'); end
+
+    # Returns the correlation_id for the message publication. Returns ''
+    # by default.
+    #
+    # Available as a method so it can be overriden (the return value could
+    # depend on the @workitem or other factors).
+    #
+    def correlation_id
+
+      opt('correlation_id') || ''
+    end
+
+    # Returns something true-ish if the participant should not reply to the
+    # engine once the publish operation is done.
+    #
+    # Available as a method so it can be overriden (the return value could
+    # depend on the @workitem or other factors).
+    #
+    def forget; opt('forget'); end
+
+    protected
+
+    # How a workitem is turned into an AMQP message payload (string).
+    #
+    # Feel free to override this method to accommodate your needs.
+    #
+    def encode_workitem
+
+      workitem.as_json
+    end
+
+    # How a "cancelitem" is turned into an AMQP message payload (string).
+    #
+    # Feel free to override this method to accommodate your needs.
+    #
+    def encode_cancelitem
+
+      Rufus::Json.encode(
+        'cancel' => true, 'fei' => @fei.h, 'flavour' => @flavour)
+    end
+
+    # Default AMQP.connect method.
+    #
+    # Feel free to override this method to accommodate your needs (See
+    # the README for an example).
+    #
+    def amqp_connect
+
+      ocon = opt('connection')
+
+      if Boss.connection && ( ! ocon)
+        Boss.connection
+      else
+        Bunny.new(Ruote.keys_to_sym(ocon || {}))
+      end
+    end
+
+    # Given connection options passed at registration time (when the
+    # participant is registered in ruote) or from the process definition,
+    # returns an AMQP::Channel instance.
+    #
+    def channel
+
+      Thread.current['_boss_amqp_channel'] ||= amqp_connect.create_channel
+    end
+
+    # Given exchange options passed at registrations time or from the process
+    # definition, returns an AMQP::Exchange instance.
+    #
+    def instantiate_exchange
+
+      Thread.current['_boss_amqp_exchange'] ||= begin
+
+        exc = exchange
+        type, name, options = exc
+
+        raise ArgumentError.new(
+          "couldn't determine exchange from #{exc.inspect}"
+        ) unless name
+
+        exchange_opts = (options || {}).each_with_object({}) { |(k, v), h|
+          h[k.to_sym] = v
+        }
+
+        Bunny::Exchange.new(channel, type.to_sym, name)#, exchange_opts)
```

Comment:
exchange_opts not used?

Reply:
No - it worked without. I don't think we use them in any way so I left the stub there as a reminder that they are intended to be passed in if something comes up in the future.

Reply:
It probably "works" without the options, but this method and module docs say that the options are passed on the exchange if they are provided, so now all that is not actually true. So either pass the options to the exchange constructor like promised, or remove the promises from the docs :)